src/Data/Singletons/Syntax.hs

```diff
@@ -132,3 +132,29 @@ buildLetDecEnv = go emptyLetDecEnv
     go acc (DInfixD f n : rest) =
       go (infixDecl f n <> acc) rest
     go acc (DPragmaD{} : rest) = go acc rest
+
+-- See Note [Standalone derived Eq instances]
+data StandaloneDerivedEqDec = SDEqDec
```

Comment:
Under my proposed refactoring, this would handle all `Eq` instances, not just standalone ones. It could then hold a `Maybe DCxt` instead of a `DCxt`.